### PR TITLE
feat(ui/image): add error slot

### PR DESCRIPTION
**Pull request by chouchouji**
## Checklist

---

- [ ] add error slot

## Change information

---

### CODE REVIEW

**Comment by haoziqaq** [comment]
需要考虑 src 发生变化时的 isError 的处理

**Comment by chouchouji** [comment]
> 需要考虑 src 发生变化时的 isError 的处理

写了一个demo，感觉大佬说的是这个意思。isError这个值我现在onload加载结束之后是没有处理的（除非是触发lazy下的errror）
https://varlet.gitee.io/varlet-ui-playground/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPSd0cyc+XG4gICBpbXBvcnQgeyByZWYsIG9uTW91bnRlZCB9IGZyb20gJ3Z1ZSdcbiAgIGNvbnN0IHVybCA9IHJlZignaHR0cHM6Ly92YXJsZXQuZ2l0ZWUuaW8vdmFybGV0LXVpLy5qcGcnKVxuICAgXG4gICBjb25zdCBoYW5kbGVFcnJvciA9ICgpID0+IHtcbiAgICAgIGNvbnNvbGUubG9nKCflm77niYflnLDlnYDkv67mlLknKVxuICAgfVxuICAgXG4gICBjb25zdCBoYW5kbGVMb2FkID0gKCkgPT4ge1xuICAgICBjb25zb2xlLmxvZygn5Yqg6L295a6M5oiQJylcbiAgIH1cbiAgIFxuICAgb25Nb3VudGVkKCgpID0+IHtcbiAgICAgc2V0VGltZW91dCgoKSA9PiB7XG4gICAgICAgdXJsLnZhbHVlID0gJ2h0dHBzOi8vdmFybGV0LmdpdGVlLmlvL3ZhcmxldC11aS9jYXQuanBnJ1xuICAgICB9LDMwMDApXG4gICB9KVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgXG4gIDxpbWcgOnNyYz1cInVybFwiIHdpZHRoPVwiNDAwXCIgaGVpZ2h0PVwiNDAwXCIgQGVycm9yPVwiaGFuZGxlRXJyb3JcIiBcbiAgICAgICBAbG9hZD1cImhhbmRsZUxvYWRcIi8+XG4gIFxuPC90ZW1wbGF0ZT5cbiIsIlBsYXlncm91bmQudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCBBcHAgZnJvbSAnLi9BcHAudnVlJ1xuaW1wb3J0IHsgaW5zdGFsbFZhcmxldFVJIH0gZnJvbSAnLi92YXJsZXQtcmVwbC1wbHVnaW4uanMnXG5cbmluc3RhbGxWYXJsZXRVSSgpXG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8QXBwIC8+XG48L3RlbXBsYXRlPlxuIiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge1xuICAgIFwidnVlXCI6IFwiaHR0cHM6Ly91bnBrZy5jb20vQHZ1ZS9ydW50aW1lLWRvbUAzLjMuMi9kaXN0L3J1bnRpbWUtZG9tLmVzbS1icm93c2VyLmpzXCIsXG4gICAgXCJAdmFybGV0L3VpXCI6IFwiLi92YXJsZXQuZXNtLmpzXCIsXG4gICAgXCJAdmFybGV0L3RvdWNoLWVtdWxhdG9yXCI6IFwiLi92YXJsZXQtdG91Y2gtZW11bGF0b3IuanNcIixcbiAgICBcIkB2YXJsZXQvdWkvanNvbi9hcmVhLmpzb25cIjogXCIuL3ZhcmxldC1hcmVhLmpzXCJcbiAgfVxufSIsInZhcmxldC1yZXBsLXBsdWdpbi5qcyI6ImltcG9ydCBWYXJsZXRVSSwgeyBDb250ZXh0IH0gZnJvbSAnQHZhcmxldC91aSdcbmltcG9ydCAnQHZhcmxldC90b3VjaC1lbXVsYXRvcidcbmltcG9ydCB7IGdldEN1cnJlbnRJbnN0YW5jZSB9IGZyb20gJ3Z1ZSdcblxuYXdhaXQgYXBwZW5kU3R5bGUoKVxuXG5leHBvcnQgZnVuY3Rpb24gaW5zdGFsbFZhcmxldFVJKCkge1xuICBjb25zdCB7IHBhcmVudCB9ID0gd2luZG93XG5cbiAgY29uc3Qgc3R5bGUgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCdzdHlsZScpXG4gIHN0eWxlLmlubmVySFRNTCA9IGBcbiAgICBib2R5IHtcbiAgICAgIG1pbi1oZWlnaHQ6IDEwMHZoO1xuICAgICAgcGFkZGluZzogMTZweDtcbiAgICAgIG1hcmdpbjogMDtcbiAgICAgIGNvbG9yOiB2YXIoLS1jb2xvci10ZXh0KTtcbiAgICAgIGJhY2tncm91bmQtY29sb3I6IHZhcigtLWNvbG9yLWJvZHkpO1xuICAgIH1cblxuICAgICo6Oi13ZWJraXQtc2Nyb2xsYmFyIHtcbiAgICAgIGRpc3BsYXk6IG5vbmU7XG4gICAgfVxuICBgXG4gIGRvY3VtZW50LmhlYWQuYXBwZW5kQ2hpbGQoc3R5bGUpXG5cbiAgaWYgKHBhcmVudC5kb2N1bWVudC5kb2N1bWVudEVsZW1lbnQuY2xhc3NMaXN0LmNvbnRhaW5zKCdkYXJrJykpIHtcbiAgICBWYXJsZXRVSS5TdHlsZVByb3ZpZGVyKFZhcmxldFVJLlRoZW1lcy5kYXJrKVxuICB9XG5cbiAgd2luZG93LmFkZEV2ZW50TGlzdGVuZXIoJ21lc3NhZ2UnLCAoeyBkYXRhIH0pID0+IHtcbiAgICBpZiAoZGF0YS5hY3Rpb24gPT09ICd0aGVtZS1jaGFuZ2UnKSB7XG4gICAgICBWYXJsZXRVSS5TdHlsZVByb3ZpZGVyKGRhdGEudmFsdWUgPT09ICdkYXJrJyA/IFZhcmxldFVJLlRoZW1lcy5kYXJrIDogbnVsbClcbiAgICB9XG4gIH0pXG5cbiAgY29uc3QgaW5zdGFuY2UgPSBnZXRDdXJyZW50SW5zdGFuY2UoKVxuICBpbnN0YW5jZS5hcHBDb250ZXh0LmFwcC51c2UoVmFybGV0VUkpXG59XG5cbmV4cG9ydCBmdW5jdGlvbiBhcHBlbmRTdHlsZSgpIHtcbiAgcmV0dXJuIG5ldyBQcm9taXNlKChyZXNvbHZlLCByZWplY3QpID0+IHtcbiAgICBjb25zdCBsaW5rID0gZG9jdW1lbnQuY3JlYXRlRWxlbWVudCgnbGluaycpXG4gICAgbGluay5yZWwgPSAnc3R5bGVzaGVldCdcbiAgICBsaW5rLmhyZWYgPSAnLi92YXJsZXQuY3NzJ1xuICAgIGxpbmsub25sb2FkID0gcmVzb2x2ZVxuICAgIGxpbmsub25lcnJvciA9IHJlamVjdFxuICAgIGRvY3VtZW50LmJvZHkuYXBwZW5kQ2hpbGQobGluaylcbiAgfSlcbn1cbiJ9

**Comment by chouchouji** [comment]
> 需要考虑 src 发生变化时的 isError 的处理

写了一个demo，感觉大佬说的是这个意思。isError这个值我现在onload加载结束之后是没有处理的（除非是触发lazy下的errror）
https://varlet.gitee.io/varlet-ui-playground/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPSd0cyc+XG4gICBpbXBvcnQgeyByZWYsIG9uTW91bnRlZCB9IGZyb20gJ3Z1ZSdcbiAgIGNvbnN0IHVybCA9IHJlZignaHR0cHM6Ly92YXJsZXQuZ2l0ZWUuaW8vdmFybGV0LXVpLy5qcGcnKVxuICAgXG4gICBjb25zdCBoYW5kbGVFcnJvciA9ICgpID0+IHtcbiAgICAgIGNvbnNvbGUubG9nKCflm77niYflnLDlnYDkv67mlLknKVxuICAgfVxuICAgXG4gICBjb25zdCBoYW5kbGVMb2FkID0gKCkgPT4ge1xuICAgICBjb25zb2xlLmxvZygn5Yqg6L295a6M5oiQJylcbiAgIH1cbiAgIFxuICAgb25Nb3VudGVkKCgpID0+IHtcbiAgICAgc2V0VGltZW91dCgoKSA9PiB7XG4gICAgICAgdXJsLnZhbHVlID0gJ2h0dHBzOi8vdmFybGV0LmdpdGVlLmlvL3ZhcmxldC11aS9jYXQuanBnJ1xuICAgICB9LDMwMDApXG4gICB9KVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgXG4gIDxpbWcgOnNyYz1cInVybFwiIHdpZHRoPVwiNDAwXCIgaGVpZ2h0PVwiNDAwXCIgQGVycm9yPVwiaGFuZGxlRXJyb3JcIiBcbiAgICAgICBAbG9hZD1cImhhbmRsZUxvYWRcIi8+XG4gIFxuPC90ZW1wbGF0ZT5cbiIsIlBsYXlncm91bmQudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCBBcHAgZnJvbSAnLi9BcHAudnVlJ1xuaW1wb3J0IHsgaW5zdGFsbFZhcmxldFVJIH0gZnJvbSAnLi92YXJsZXQtcmVwbC1wbHVnaW4uanMnXG5cbmluc3RhbGxWYXJsZXRVSSgpXG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8QXBwIC8+XG48L3RlbXBsYXRlPlxuIiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge1xuICAgIFwidnVlXCI6IFwiaHR0cHM6Ly91bnBrZy5jb20vQHZ1ZS9ydW50aW1lLWRvbUAzLjMuMi9kaXN0L3J1bnRpbWUtZG9tLmVzbS1icm93c2VyLmpzXCIsXG4gICAgXCJAdmFybGV0L3VpXCI6IFwiLi92YXJsZXQuZXNtLmpzXCIsXG4gICAgXCJAdmFybGV0L3RvdWNoLWVtdWxhdG9yXCI6IFwiLi92YXJsZXQtdG91Y2gtZW11bGF0b3IuanNcIixcbiAgICBcIkB2YXJsZXQvdWkvanNvbi9hcmVhLmpzb25cIjogXCIuL3ZhcmxldC1hcmVhLmpzXCJcbiAgfVxufSIsInZhcmxldC1yZXBsLXBsdWdpbi5qcyI6ImltcG9ydCBWYXJsZXRVSSwgeyBDb250ZXh0IH0gZnJvbSAnQHZhcmxldC91aSdcbmltcG9ydCAnQHZhcmxldC90b3VjaC1lbXVsYXRvcidcbmltcG9ydCB7IGdldEN1cnJlbnRJbnN0YW5jZSB9IGZyb20gJ3Z1ZSdcblxuYXdhaXQgYXBwZW5kU3R5bGUoKVxuXG5leHBvcnQgZnVuY3Rpb24gaW5zdGFsbFZhcmxldFVJKCkge1xuICBjb25zdCB7IHBhcmVudCB9ID0gd2luZG93XG5cbiAgY29uc3Qgc3R5bGUgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCdzdHlsZScpXG4gIHN0eWxlLmlubmVySFRNTCA9IGBcbiAgICBib2R5IHtcbiAgICAgIG1pbi1oZWlnaHQ6IDEwMHZoO1xuICAgICAgcGFkZGluZzogMTZweDtcbiAgICAgIG1hcmdpbjogMDtcbiAgICAgIGNvbG9yOiB2YXIoLS1jb2xvci10ZXh0KTtcbiAgICAgIGJhY2tncm91bmQtY29sb3I6IHZhcigtLWNvbG9yLWJvZHkpO1xuICAgIH1cblxuICAgICo6Oi13ZWJraXQtc2Nyb2xsYmFyIHtcbiAgICAgIGRpc3BsYXk6IG5vbmU7XG4gICAgfVxuICBgXG4gIGRvY3VtZW50LmhlYWQuYXBwZW5kQ2hpbGQoc3R5bGUpXG5cbiAgaWYgKHBhcmVudC5kb2N1bWVudC5kb2N1bWVudEVsZW1lbnQuY2xhc3NMaXN0LmNvbnRhaW5zKCdkYXJrJykpIHtcbiAgICBWYXJsZXRVSS5TdHlsZVByb3ZpZGVyKFZhcmxldFVJLlRoZW1lcy5kYXJrKVxuICB9XG5cbiAgd2luZG93LmFkZEV2ZW50TGlzdGVuZXIoJ21lc3NhZ2UnLCAoeyBkYXRhIH0pID0+IHtcbiAgICBpZiAoZGF0YS5hY3Rpb24gPT09ICd0aGVtZS1jaGFuZ2UnKSB7XG4gICAgICBWYXJsZXRVSS5TdHlsZVByb3ZpZGVyKGRhdGEudmFsdWUgPT09ICdkYXJrJyA/IFZhcmxldFVJLlRoZW1lcy5kYXJrIDogbnVsbClcbiAgICB9XG4gIH0pXG5cbiAgY29uc3QgaW5zdGFuY2UgPSBnZXRDdXJyZW50SW5zdGFuY2UoKVxuICBpbnN0YW5jZS5hcHBDb250ZXh0LmFwcC51c2UoVmFybGV0VUkpXG59XG5cbmV4cG9ydCBmdW5jdGlvbiBhcHBlbmRTdHlsZSgpIHtcbiAgcmV0dXJuIG5ldyBQcm9taXNlKChyZXNvbHZlLCByZWplY3QpID0+IHtcbiAgICBjb25zdCBsaW5rID0gZG9jdW1lbnQuY3JlYXRlRWxlbWVudCgnbGluaycpXG4gICAgbGluay5yZWwgPSAnc3R5bGVzaGVldCdcbiAgICBsaW5rLmhyZWYgPSAnLi92YXJsZXQuY3NzJ1xuICAgIGxpbmsub25sb2FkID0gcmVzb2x2ZVxuICAgIGxpbmsub25lcnJvciA9IHJlamVjdFxuICAgIGRvY3VtZW50LmJvZHkuYXBwZW5kQ2hpbGQobGluaylcbiAgfSlcbn1cbiJ9

File: packages/varlet-ui/src/image/Image.vue [hunk 1]
```diff
@@ -64,12 +64,16 @@ export default defineComponent({
...
+          call(onLoad, e)
+        }
         if (el._lazy.state === 'error') {
-          call(onError, e)
           isError.value = true
+          call(onError, e)
         }
       } else {
+        isError.value = false
         call(onLoad, e)
```
Comment (haoziqaq) [inline packages/varlet-ui/src/image/Image.vue]: 现在的处理逻辑有可能会产生以下情况:

lazy 模式下，在图片传递了 error 插槽并且 isError.value = true 之后（此时页面上应该显示的是错误插槽），改变了 src, 这时应该进入 loading，但是目前的逻辑可能不会正确的显示出 loading 图片。

我建议直接 watch src，然后将 isError.value = false 进行重置，可以尝试一下是否可行。

Comment (chouchouji) [reply packages/varlet-ui/src/image/Image.vue]: 晚会时候我试一下

Comment (chouchouji) [reply packages/varlet-ui/src/image/Image.vue]: 大佬说的对， lazy 模式下 由src改变导致的 error -> success 转变的过程中，虽然 src 的值成功改了，但是图片并没有预期那样显示出来，没有触发 handleLoad 事件


https://github.com/varletjs/varlet/assets/70570907/7f71f5f4-903d-41a6-870f-e5cc8d7b7714

截屏2023-05-30 09 12 25

截屏2023-05-30 09 12 41

Comment (chouchouji) [reply packages/varlet-ui/src/image/Image.vue]: 按照大佬说的那种方式修改之后，`image` 组件行为都是符合预期的👍。我还测试了一下极端的情况，根据按钮来回切换 src 的值，也正常。原来我的写法不合理之处应该是 `lazy` 模式下，`onload` 事件只有在实际加载图像并将其添加到文档后才会触发，改成监听之后，每次修改 `src` 的值，`v-if` 的值的变化牵扯到组件的销毁和创建，所以能正常触发 `onload` 等事件。大佬能看下我这里的理解是不是正确的嘛，阿里嘎多

File: packages/varlet-ui/src/image/docs/zh-CN.md [hunk 2]
```diff
@@ -121,6 +121,20 @@
 </template>
 ```
 
+### 加载失败提示
```
Comment (haoziqaq) [inline packages/varlet-ui/src/image/docs/zh-CN.md:124]: 修改为 `加载失败插槽`

File: packages/varlet-ui/src/image/docs/en-US.md [hunk 3]
```diff
@@ -204,3 +217,9 @@ import logo from '../../assets/logo.png'
 | `click` | Triggered when you click on Image | `event: Event` |
 | `load` | Triggered when the image is successfully loaded (when the image fails in lazy loading mode, it will only be triggered when the image is successfully loaded) | `event: Event` |
 | `error` | Triggered when an image fails to load (when an image fails in lazy loading mode, it will only be triggered when the number of attempts ends) | `event: Event` |
+
+### Slot
```
Comment (haoziqaq) [inline packages/varlet-ui/src/image/docs/en-US.md:221]: Slots，这里涉及到 extension 的信息编译，所以一定得是 Slots

File: packages/varlet-ui/src/image/docs/zh-CN.md [hunk 4]
```diff
@@ -121,6 +121,20 @@
...
+
+```html
+<template>
+   <var-image src="https://varlet.gitee.io/varlet-ui/ca.jpg">
```
Comment (haoziqaq) [inline packages/varlet-ui/src/image/docs/zh-CN.md:128]: https://varlet.gitee.io/varlet-ui/cat.jpg

style 内联样式太长了，修改成 class

File: packages/varlet-ui/src/image/docs/zh-CN.md [hunk 5]
```diff
@@ -203,3 +216,9 @@ import logo from '../../assets/logo.png'
 | `click` | 点击图片时触发 | `event: Event` |
 | `load` | 图片成功加载时触发(懒加载模式下失败时会多次尝试加载图片，只会在成功加载时触发) | `event: Event` |
 | `error` | 图片失败加载时触发(懒加载模式下失败时会多次尝试加载图片，只会在尝试次数结束时触发) | `event: Event` |
+
+### 插槽
+
+| 插槽名 | 说明 | 参数 |
+| --- | --- | --- |
+| `error` | 自定义加载失败后的提示内容 | `-` |
```
Comment (haoziqaq) [inline packages/varlet-ui/src/image/docs/zh-CN.md:224]: 改为加载失败插槽

File: packages/varlet-ui/src/image/Image.vue [hunk 6]
```diff
@@ -12,10 +12,9 @@
       :class="n('image')"
       :alt="alt"
       :title="title"
-      :lazy-error="error"
```
Comment (haoziqaq) [inline packages/varlet-ui/src/image/Image.vue:15]: 这个 api 不应该被删除，他应该和插槽兼容。

**Review by haoziqaq — COMMENTED** [review]
在 types 里应该完成 error 插槽的类型

File: packages/varlet-ui/src/image/docs/en-US.md (hunk 3 above)
Comment (chouchouji) [reply packages/varlet-ui/src/image/docs/en-US.md:221]: ok，俺记住了

**Pull request merged**